### PR TITLE
[Routing] Defaults by Symfony Config

### PR DESCRIPTION
**Pull request by blankse**
We want use the constants to set a default bundle, controller and view. In the moment it is not possible.

### CODE REVIEW

**Comment by dpfaffenbauer** [comment]
Please don't introduce anymore constants, use Container Parameters instead, you can fill them with constant values or env vars.

**Comment by blankse** [comment]
These are no new constants. Are documented in the example:
https://github.com/pimcore/skeleton/blob/master/app/constants.example.php

But yes should be parameters...

**Comment by blankse** [comment]
We could integrate these settings in the symfony configuration tree and set the default value to these constants. 
https://github.com/pimcore/pimcore/blob/master/bundles/CoreBundle/DependencyInjection/Configuration.php

WDYT? Any suggestions for the config tree?

**Comment by dpfaffenbauer** [comment]
oh wow, true, these contants really exists. But as they don't work anyway, I would much rather use the Symfony Config Tree for that

**Comment by brusch** [comment]
Config tree would be fine for me ... what about adding them to the existing `routing` node? 
```yml
pimcore: 
  routing: 
    defaults: 
      bundle: AppBundle
      controller: Default
      action: default
```

**Comment by dpfaffenbauer** [comment]
👍

**Comment by brusch** [comment]
@blankse can you update your PR, please?  :)

**Comment by blankse** [comment]
@brusch Yes I have already started :)

**Comment by blankse** [comment]
@brusch I updated the PR :)

File: bundles/CoreBundle/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -350,6 +350,20 @@ private function addRoutingNode(ArrayNodeDefinition $rootNode)
                 ->arrayNode('routing')
                     ->addDefaultsIfNotSet()
                     ->children()
+                        ->arrayNode('defaults')
+                            ->addDefaultsIfNotSet()
+                            ->children()
+                                ->scalarNode('bundle')
+                                    ->defaultValue(PIMCORE_SYMFONY_DEFAULT_BUNDLE)
```
Comment (brusch) [inline bundles/CoreBundle/DependencyInjection/Configuration.php]: Imho we can completely get rid of those constants and define the default values in the config-tree definition. 
If someone wants to use constants anyway, there's a special syntax for the use in the config.

Comment (blankse) [reply bundles/CoreBundle/DependencyInjection/Configuration.php]: I removed the constants and add an BC notice

**Comment by blankse** [comment]
I added docs. There I noticed that it concerns only the routing of the documents. Should not it be in the documents tree?

**Comment by brusch** [comment]
No, it's not only used for documents, but also for generic purposes, like here: 
https://github.com/pimcore/pimcore/blob/4c986d60d890898e466533a21dfd29bde2acb9f6/bundles/CoreBundle/EventListener/TemplateControllerListener.php#L110 
or for static routes, here: 
https://github.com/pimcore/pimcore/blob/433957d8aaf67d4054b0c2bcb918ee48e8c78a10/lib/Tool.php#L299

We could think about a dedicated config for documents, but that sounds a bit too much for me :)

**Comment by dpfaffenbauer** [comment]
There are already pre-defined document types, so that should be enough

**Comment by blankse** [comment]
Rebase because of merge conflict in README.md

**Comment by brusch** [comment]
Hmm, but with your last changes you made the system setting for the default controller & action obsolete - which is fine for me, but then we should also remove it from the UI and announce that in the upgrade notes.

**Comment by brusch** [comment]
Sorry, pressed the wrong button 🤓

**Comment by blankse** [comment]
@brusch I removed the system settings for default controller and action and add the bc break notice

**Comment by blankse** [comment]
@brusch Did you mean that the settings were no longer used? The Tool:getRoutingDefaults() function had used it.

**Comment by brusch** [comment]
@blankse oh yes, you're absolutely right ... but the system settings had priority, so the 2 config options for action and controller were pretty much useless ... especially for existing setups where the default controller/action are set in system settings. I'd say the current implementation is more consistent and less confusing 👌

File: bundles/AdminBundle/Controller/Admin/MiscController.php [hunk 2]
```diff
@@ -93,12 +94,17 @@ public function getAvailableControllersAction(Request $request, ControllerDataPr
...
+        $routingDefaults = $this->getParameter('pimcore.routing.defaults');
         $bundle = $request->get('moduleName');
         if (empty($bundle)) {
-            $bundle = 'AppBundle';
+            $bundle = $routingDefaults['bundle'];
         }
 
         $controller = $request->get('controllerName');
+        if (empty($controller)) {
+            $bundle = $routingDefaults['controller'];
```
Comment (brusch) [inline bundles/AdminBundle/Controller/Admin/MiscController.php]: This should be `$controller = ...` i guess.

Comment (blankse) [reply bundles/AdminBundle/Controller/Admin/MiscController.php]: Of course, I fixed it

**Comment by brusch** [comment]
I guess that's it ... let's wait for the tests to be finished, then I'm going to merge the PR.

File: lib/Tool.php [hunk 3]
```diff
@@ -289,29 +289,11 @@ public static function getLanguageFlagFile($language)
...
-
-            // get configured settings for defaults
-            $systemRoutingDefaults = $config->documents->toArray();
-
-            foreach ($routingDefaults as $key => $value) {
-                if (isset($systemRoutingDefaults['default_' . $key]) && $systemRoutingDefaults['default_' . $key]) {
-                    $routingDefaults[$key] = $systemRoutingDefaults['default_' . $key];
-                }
-            }
+        $container = \Pimcore::getContainer();
```
Comment (dpfaffenbauer) [inline lib/Tool.php]: you could move that to a compiler pass, so there would be one less call to \Pimcore::getContainer() ;)

Comment (blankse) [reply lib/Tool.php]: You mean the module-bundle mapping? I think we can do this in the extension class too. Than we can use everywhere the paramerter directly and can remove this function. Is this what you think?

Comment (dpfaffenbauer) [reply lib/Tool.php]: Thinking about that: I guess you don't really need the getRoutingDefaults function at all. You can load it from the container directly when you need it. The only place you would need to call ```Pimcore::getContainer``` again is the Renderlet, cause they don't support dependency injection.

Comment (brusch) [reply lib/Tool.php]: For me it's fine as it is now, I'd keep `getRoutingDefaults()` anyway to avoid possible BC issues.

**Pull request merged**